presto-native-execution/presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporter.cpp

```diff
@@ -0,0 +1,222 @@
+  if (pcts.size() > 0) {
+    auto& summaryFamily = BuildSummary()
+                              .Name(keyStr.append(kSummarySuffix))
+                              .Register(*impl_->registry);
+    Summary::Quantiles quantiles;
+    for (auto pct : pcts) {
+      quantiles.push_back(
+          detail::CKMSQuantiles::Quantile(pct / (double)100, 0));
+    }
+    auto& summaryMetric = summaryFamily.Add({impl_->labels}, quantiles);
+    registeredMetricsMap_.emplace(
+        std::string(key).append(kSummarySuffix),
+        StatsInfo{velox::StatType::COUNT, &summaryMetric});
+  }
+}
+
+void PrometheusReporter::registerHistogramMetricExportType(
+    folly::StringPiece key,
+    int64_t bucketWidth,
+    int64_t min,
+    int64_t max,
+    const std::vector<int32_t>& pcts) const {
+  registerHistogramMetricExportType(
+      key.toString().c_str(), bucketWidth, min, max, pcts);
+}
+
+void PrometheusReporter::addMetricValue(const std::string& key, size_t value)
+    const {
+  addMetricValue(key.c_str(), value);
+}
+
+void PrometheusReporter::addMetricValue(const char* key, size_t value) const {
+  auto statsInfo = registeredMetricsMap_.find(key)->second;
+  switch (statsInfo.statType) {
+    case velox::StatType::COUNT: {
+      Counter* counter = (Counter*)(statsInfo.metricPtr);
+      counter->Increment(value);
+    } break;
+    case velox::StatType::SUM:
+    case velox::StatType::AVG:
+    case velox::StatType::RATE: {
+      // Overrides the existing state.
+      Gauge* gauge = (Gauge*)(statsInfo.metricPtr);
+      gauge->Set(value);
+    } break;
+  };
+}
+
+void PrometheusReporter::addMetricValue(folly::StringPiece key, size_t value)
+    const {
+  addMetricValue(key.toString().c_str(), value);
+}
+
+void PrometheusReporter::addHistogramMetricValue(
+    const std::string& key,
+    size_t value) const {
+  addHistogramMetricValue(key.c_str(), value);
+}
+
+void PrometheusReporter::addHistogramMetricValue(const char* key, size_t value)
+    const {
+  if (!registeredMetricsMap_.count(key)) {
+    VLOG(1) << "addMetricValue for unregistered metric " << key;
+    return;
+  }
+  auto* histgoram =
+      (Histogram*)registeredMetricsMap_.find(key)->second.metricPtr;
+  histgoram->Observe(value);
+  std::string summaryKey = std::string(key).append(kSummarySuffix);
+  if (registeredMetricsMap_.count(summaryKey)) {
+    auto* summary =
+        (Summary*)registeredMetricsMap_.find(summaryKey)->second.metricPtr;
+    summary->Observe(value);
+  }
+}
+
+void PrometheusReporter::addHistogramMetricValue(
+    folly::StringPiece key,
+    size_t value) const {
+  addHistogramMetricValue(key.toString().c_str(), value);
+}
+
+std::string PrometheusReporter::getSerializedMetrics() const {
+  if (registeredMetricsMap_.empty()) {
+    return "";
+  }
+  TextSerializer serializer;
+  // Registry::Collect() acquires lock on a mutex.
+  return serializer.Serialize(impl_->registry->Collect());
+}
+
+velox::BaseStatsReporter* createPrometheusReporter() {
```

Comment:
nit: From an API standpoint, we can return a `unique_ptr` here and release the ownership if needed inside 
`folly::Singleton<facebook::velox::BaseStatsReporter> reporter`

Reply:
If you call unique_ptr.release(), the ownership will be released and the destructor should not be invoked.